# Customer analysis

In this exercise, you work for a consulting firm as data analyst. One of your clients asked for help to understand its customer buying behavior. 

You dispose of a dataset from its e-commerce website with 13 features : 

- order_id : ID if the order
- order_status : Status of the order
- order_purchase_timestamp : Timestamp of the purchase
- order_delivered_customer_date : Timestamp of the delivery to the customer
- payment_type : Method of payment
- payment_installments : Number of payments for this purchase (does NOT affect the payment value)
- payment_value : Total amount of the payment
- review_score : Score given by the client, ranges from 1 to 5
- customer_unique_id : ID of the customer who purchased
- customer_zip_code_prefix : Zip code of the customer
- customer_city : City of the customer
- customer_state : State of the customer
- product_category_name_english : Name of the product category in English

With this data, you have to analyze customer behavior with two techniques:

- Customer segmentation with RFM
- Cohort analysis from Sept 2016 to October 2018.

At the end, you have to present your analysis with actionable insights to improve the customer relationship of your client. 

# Understanding RFM Segmentation 

❓ What are the 3 metrics used in an RFM Segmentation? (hint : What does RFM stand for?) 

❓ Can you summarize what each metric measures in one sentence?

❓ What data / or type of data do you need for each metric ? 

Let's explore the dataset!

# Exploratory Data Analysis

## Import

Load the dataset and start an Exploratory Data Analysis

In [9]:
# Libraries
import pandas as pd
from datetime import datetime, timedelta
import numpy as np
from operator import attrgetter
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors

In [10]:
df = pd.read_csv('data/db_customers.csv')

In [4]:
df

Unnamed: 0                          order_id  \
0                0  e481f51cbdc54678b7cc49136f2d6af7   
1                1  e481f51cbdc54678b7cc49136f2d6af7   
2                2  e481f51cbdc54678b7cc49136f2d6af7   
3                3  53cdb2fc8bc7dce0b6741e2150273451   
4                4  47770eb9100c2d0c44946d9cf07ec65d   
...            ...                               ...   
105378      105378  9c5dedf39a927c1b2549525ed64a053c   
105379      105379  63943bddc261676b46f01ca7ac2f7bd8   
105380      105380  83c1379a015df1e13d02aae0204711ab   
105381      105381  11c177c8e97725db2631073c19f07b62   
105382      105382  66dea50a8b16d9b4dee7af250b4be1a5   

                             customer_id order_status  \
0       9ef432eb6251297304e76186b10a928d    delivered   
1       9ef432eb6251297304e76186b10a928d    delivered   
2       9ef432eb6251297304e76186b10a928d    delivered   
3       b0830fb4747a6c6d20dea0b8c802d7ef    delivered   
4       41ce2a54c0b03bf3443c3d931a367089    delivered   
...                                  ...          ...   
105378  39bd1228ee8140590ac3aca26f2dfe00    delivered   
105379  1fca14ff2861355f6e5f14306ff977a7    delivered   
105380  1aa71eb042121263aafbe80c1b562c9c    delivered   
105381  b331b74b18dc79bcdf6532d51e1637c1    delivered   
105382  edb027a75a1449115f6b43211ae02a24    delivered   

       order_purchase_timestamp order_delivered_customer_date payment_type  \
0           2017-10-02 10:56:33           2017-10-10 21:25:13  credit_card   
1           2017-10-02 10:56:33           2017-10-10 21:25:13      voucher   
2           2017-10-02 10:56:33           2017-10-10 21:25:13      voucher   
3           2018-07-24 20:41:37           2018-08-07 15:27:45       boleto   
4           2018-08-08 08:38:49           2018-08-17 18:06:29  credit_card   
...                         ...                           ...          ...   
105378      2017-03-09 09:54:05           2017-03-17 15:08:01  credit_card   
105379      2018-02-06 12:58:58           2018-02-28 17:37:56  credit_card   
105380      2017-08-27 14:46:43           2017-09-21 11:24:17  credit_card   
105381      2018-01-08 21:28:27           2018-01-25 23:32:54  credit_card   
105382      2018-03-08 20:57:30           2018-03-16 13:08:30   debit_card   

        payment_installments  payment_value  review_score  \
0                        1.0          18.12             4   
1                        1.0           2.00             4   
2                        1.0          18.59             4   
3                        1.0         141.46             4   
4                        3.0         179.12             5   
...                      ...            ...           ...   
105378                   3.0          85.08             5   
105379                   3.0         195.00             4   
105380                   5.0         271.01             5   
105381                   4.0         441.16             2   
105382                   1.0          86.86             5   

                      customer_unique_id  customer_zip_code_prefix  \
0       7c396fd4830fd04220f754e42b4e5bff                      3149   
1       7c396fd4830fd04220f754e42b4e5bff                      3149   
2       7c396fd4830fd04220f754e42b4e5bff                      3149   
3       af07308b275d755c9edb36a90c618231                     47813   
4       3a653a41f6f9fc3d2a113cf8398680e8                     75265   
...                                  ...                       ...   
105378  6359f309b166b0196dbf7ad2ac62bb5a                     12209   
105379  da62f9e57a76d978d02ab5362c509660                     11722   
105380  737520a9aad80b3fbbdad19b66b37b30                     45920   
105381  5097a5312c8b157bb7be58ae360ef43c                     28685   
105382  60350aa974b26ff12caad89e55993bd6                     83750   

              customer_city customer_state product_category_name_english  
0                 sao paulo             SP                    ho

❓ Which features should be formated in datetime ? 
Convert them

In [13]:
df[['order_purchase_timestamp', 'order_delivered_customer_date']] = df[['order_purchase_timestamp', 'order_delivered_customer_date']].apply(pd.to_datetime)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105383 entries, 0 to 105382
Data columns (total 15 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   Unnamed: 0                     105383 non-null  int64         
 1   order_id                       105383 non-null  object        
 2   customer_id                    105383 non-null  object        
 3   order_status                   105383 non-null  object        
 4   order_purchase_timestamp       105383 non-null  datetime64[ns]
 5   order_delivered_customer_date  102224 non-null  datetime64[ns]
 6   payment_type                   105382 non-null  object        
 7   payment_installments           105382 non-null  float64       
 8   payment_value                  105382 non-null  float64       
 9   review_score                   105383 non-null  int64         
 10  customer_unique_id             105383 non-null  object        
 11  

In [14]:
df

Unnamed: 0                          order_id  \
0                0  e481f51cbdc54678b7cc49136f2d6af7   
1                1  e481f51cbdc54678b7cc49136f2d6af7   
2                2  e481f51cbdc54678b7cc49136f2d6af7   
3                3  53cdb2fc8bc7dce0b6741e2150273451   
4                4  47770eb9100c2d0c44946d9cf07ec65d   
...            ...                               ...   
105378      105378  9c5dedf39a927c1b2549525ed64a053c   
105379      105379  63943bddc261676b46f01ca7ac2f7bd8   
105380      105380  83c1379a015df1e13d02aae0204711ab   
105381      105381  11c177c8e97725db2631073c19f07b62   
105382      105382  66dea50a8b16d9b4dee7af250b4be1a5   

                             customer_id order_status  \
0       9ef432eb6251297304e76186b10a928d    delivered   
1       9ef432eb6251297304e76186b10a928d    delivered   
2       9ef432eb6251297304e76186b10a928d    delivered   
3       b0830fb4747a6c6d20dea0b8c802d7ef    delivered   
4       41ce2a54c0b03bf3443c3d931a367089    delivered   
...                                  ...          ...   
105378  39bd1228ee8140590ac3aca26f2dfe00    delivered   
105379  1fca14ff2861355f6e5f14306ff977a7    delivered   
105380  1aa71eb042121263aafbe80c1b562c9c    delivered   
105381  b331b74b18dc79bcdf6532d51e1637c1    delivered   
105382  edb027a75a1449115f6b43211ae02a24    delivered   

       order_purchase_timestamp order_delivered_customer_date payment_type  \
0           2017-10-02 10:56:33           2017-10-10 21:25:13  credit_card   
1           2017-10-02 10:56:33           2017-10-10 21:25:13      voucher   
2           2017-10-02 10:56:33           2017-10-10 21:25:13      voucher   
3           2018-07-24 20:41:37           2018-08-07 15:27:45       boleto   
4           2018-08-08 08:38:49           2018-08-17 18:06:29  credit_card   
...                         ...                           ...          ...   
105378      2017-03-09 09:54:05           2017-03-17 15:08:01  credit_card   
105379      2018-02-06 12:58:58           2018-02-28 17:37:56  credit_card   
105380      2017-08-27 14:46:43           2017-09-21 11:24:17  credit_card   
105381      2018-01-08 21:28:27           2018-01-25 23:32:54  credit_card   
105382      2018-03-08 20:57:30           2018-03-16 13:08:30   debit_card   

        payment_installments  payment_value  review_score  \
0                        1.0          18.12             4   
1                        1.0           2.00             4   
2                        1.0          18.59             4   
3                        1.0         141.46             4   
4                        3.0         179.12             5   
...                      ...            ...           ...   
105378                   3.0          85.08             5   
105379                   3.0         195.00             4   
105380                   5.0         271.01             5   
105381                   4.0         441.16             2   
105382                   1.0          86.86             5   

                      customer_unique_id  customer_zip_code_prefix  \
0       7c396fd4830fd04220f754e42b4e5bff                      3149   
1       7c396fd4830fd04220f754e42b4e5bff                      3149   
2       7c396fd4830fd04220f754e42b4e5bff                      3149   
3       af07308b275d755c9edb36a90c618231                     47813   
4       3a653a41f6f9fc3d2a113cf8398680e8                     75265   
...                                  ...                       ...   
105378  6359f309b166b0196dbf7ad2ac62bb5a                     12209   
105379  da62f9e57a76d978d02ab5362c509660                     11722   
105380  737520a9aad80b3fbbdad19b66b37b30                     45920   
105381  5097a5312c8b157bb7be58ae360ef43c                     28685   
105382  60350aa974b26ff12caad89e55993bd6                     83750   

              customer_city customer_state product_category_name_english  
0                 sao paulo             SP                    ho

❓ Which features are Categorical ? List the unique values on these features.

These features are categorical: [order_status] | [payment_type] | [customer_city] | [customer_state] | [product_category_name_english]

In [15]:

# Storing unique values for categorical features into variables
order_status=df['order_status'].unique()
payment_type=df['payment_type'].unique()
customer_city=df['customer_city'].unique()
customer_state=df['customer_state'].unique()
product_category_name_english=df['product_category_name_english'].unique()

# Printing unique values for each categorical columns
print(f'order status: {order_status}')
print(f'payment_type: {payment_type}')
print(f'order status: {customer_city}')
print(f'order status: {customer_state}')
print(f'order status: {product_category_name_english}')

order status: ['delivered' 'invoiced' 'shipped' 'processing' 'unavailable' 'canceled'
 'created' 'approved']
payment_type: ['credit_card' 'voucher' 'boleto' 'debit_card' 'not_defined' nan]
order status: ['sao paulo' 'barreiras' 'vianopolis' ... 'messias targino'
 'campo do tenente' 'nova vicosa']
order status: ['SP' 'BA' 'GO' 'RN' 'PR' 'RS' 'RJ' 'MG' 'SC' 'RR' 'PE' 'TO' 'CE' 'DF'
 'SE' 'MT' 'PB' 'PA' 'RO' 'ES' 'AP' 'MS' 'MA' 'PI' 'AL' 'AC' 'AM']
order status: ['housewares' 'perfumery' 'auto' 'pet_shop' 'stationery' nan
 'furniture_decor' 'office_furniture' 'garden_tools'
 'computers_accessories' 'bed_bath_table' 'toys'
 'construction_tools_construction' 'telephony' 'health_beauty'
 'electronics' 'baby' 'cool_stuff' 'watches_gifts' 'air_conditioning'
 'sports_leisure' 'books_general_interest' 'small_appliances' 'food'
 'luggage_accessories' 'fashion_underwear_beach' 'christmas_supplies'
 'fashion_bags_accessories' 'musical_instruments'
 'construction_tools_lights' 'books_technical' 'cos

❓ Check for missing values. What do you notice ? 


In [17]:
# Checking for missing values
df.isnull().sum()


Unnamed: 0                          0
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_delivered_customer_date    3159
payment_type                        1
payment_installments                1
payment_value                       1
review_score                        0
customer_unique_id                  0
customer_zip_code_prefix            0
customer_city                       0
customer_state                      0
product_category_name_english    2410
dtype: int64

We can see missing values in 5 features:
1) payment_type, payement_installments, payment_value (1 missing value in each) 

2) order_delivered_customer_date (3159)

3) product_category_name_english(2410)

❓ Can you find an explanation for the missing `order_delivered_customer_date` ? 
(hint : Check for non delivered orders)

In [18]:
# Checking if missing order_delivered_customer_date count is equal to all purchases which are not yet delivered
df[df['order_status'] != "delivered"].shape

(3158, 15)

We have 3158 missing values for order_delivered_customer_date. It's because these purchases are not delivered yet. There is a difference of 1, and I think it's because of the missing value in the 3 other features (payment_type,installments,value).

❓ What are the proportions of each `order_status` ? 

In [19]:
# Calculate ratio of each order status in our dataframe.
df['order_status'].value_counts(normalize=True).mul(100).round(1).astype(str) + '%'

delivered      97.0%
shipped         1.1%
canceled        0.6%
unavailable     0.6%
invoiced        0.3%
processing      0.3%
created         0.0%
approved        0.0%
Name: order_status, dtype: object

❓ Which `order_status` do you think should be removed from the analysis

I would removed all the order_status, except delivered and shipped status. These 2 orders represent 98% of the order_status, and we want to focus on this because it will be easier to segmentate our different customers.

❓ What feature would you choose in the e-commerce dataset provided? 

❓ What method would you use to group the data? 

❓ What feature would you use to group the data? 

Group the data. (Hint: you can carry out the grouping one by one and then merge, or even better use the .agg method and get a robust grouping in one line of code)

What type of data is the R supposed to be? Is it in the right format? (Hint : investigate the difference between `datetime` and `timedelta`) 

So there is one more step to carry out before the data is ready : Calculate the recency base on the date of the purchase

❓ What is the date of the least purchase?

❓What is the date of the most recent purchase?

❓What is the date of the most recent purchase?

❓How would you calculate the recency? (hint: what date would you use to compare with the purchase date? Today? probably not...)

Create a feature calculating the difference between `the most recent purchased + 1 day`, and `order_purchase_timestamp`.

Now that you have created a new dataset with the right columns, explore it to understand what you created.

❓Which column is the R? The F? The M?

Now for each metric independantly, build clutsers (Hint 1: you can use `.describe()` to have insights about each metrics. Hint 2: Don't choose more than 3 clusters)

Now that you have clusters for each category, explain the different buying behaviors

# Working with cohorts

❓What is a cohort? 

❓Why is it interesting to analyse cohorts ? 

❓What data (already existing in the original dataset) do we need to build a cohort? 

Build a new df with just these 3 features

Using the `pandas.Serie.dt.to_period()` methode, create a new feature converting `order_purchase_timestamp` in months

Create a new feature called `cohort` by grouping the previous df by `customer_unique_id`, and tranforming it's minimum value in months (with the same `pandas.Serie.dt.to_period()` as before)

In [ ]:
# For the teachers, To filter by only the customer that have done 2 or more purchases
df_more =pd.DataFrame(df_c.groupby('customer_unique_id').size()[df_c.groupby('customer_unique_id').size() > 1])
df_c = df_c.merge(df_more, how='left', on='customer_unique_id').rename(columns={0:'na'},inplace=True)

df_c = df_c[df_c['na'].notna()]

In [ ]:
df_c['cohort'] = df_c.groupby('customer_unique_id')['order_purchase_timestamp'] \
                 .transform('min') \
                 .dt.to_period('M') 
df_c

Now groupby `['cohort', 'order_month']` and aggregate by `nunique()` `customer_unique_id`

Finally, create a feature called `['period_number']` where you substract the `cohort` to the `order_month` and get the `'n'` attribute thanks to the `operator.attrgetter` method

Explore a cohort (for example 2017-01) and try to understand what it shows

Pivot the DataFrame using the methode : `DataFrame.pivot_table` (hint : use `'cohort'` as `index`, `'period_number'` as columns, and `'n_customers'` as `values`)

Now divide all the values (except you column 0) by the column 0 and name this new DataFrame `retention_matrix`

Plot your `retention matrix` using the code below : 

In [ ]:

with sns.axes_style("white"):
    fig, ax = plt.subplots(1, 2, figsize=(12, 8), sharey=True, gridspec_kw={'width_ratios': [1, 11]})
    
    # retention matrix
    sns.heatmap(retention_matrix, 
                mask=retention_matrix.isnull(), 
                annot=True, 
                fmt='.0%', 
                cmap='RdYlGn', 
                ax=ax[1])
    ax[1].set_title('Monthly Cohorts: User Retention', fontsize=16)
    ax[1].set(xlabel='# of periods',
              ylabel='')

    # cohort size
    cohort_size_df = pd.DataFrame(cohort_size).rename(columns={0: 'cohort_size'})
    white_cmap = mcolors.ListedColormap(['white'])
    sns.heatmap(cohort_size_df, 
                annot=True, 
                cbar=False, 
                fmt='g', 
                cmap=white_cmap, 
                ax=ax[0])

    fig.tight_layout()


What do you notice? 

How could we have know that before? (hint : Calculate the percentage of customers who have ordered more than once)

Here is the code for the analysis of customers who have bought more than once : 

In [ ]:
#start from your df created doing the RFM segmentation
df_c = df[['customer_unique_id', 'order_id',
           'order_purchase_timestamp']].drop_duplicates()

df_c['order_month'] = df_c['order_purchase_timestamp'].dt.to_period('M')

df_more = pd.DataFrame(df_c.groupby('customer_unique_id').size()[
                       df_c.groupby('customer_unique_id').size() > 1])

df_c = df_c.merge(df_more, how='left', on='customer_unique_id').rename(
    columns={0: 'na'})

df_c = df_c[df_c['na'].notna()]
df_c['cohort'] = df_c.groupby('customer_unique_id')['order_purchase_timestamp'] \
    .transform('min') \
    .dt.to_period('M')

df_cohort = df_c.groupby(['cohort', 'order_month']) \
    .agg(n_customers=('customer_unique_id', 'nunique')) \
    .reset_index(drop=False)
df_cohort['period_number'] = (
    df_cohort.order_month - df_cohort.cohort).apply(attrgetter('n'))
cohort_pivot = df_cohort.pivot_table(index='cohort',
                                     columns='period_number',
                                     values='n_customers')
cohort_size = cohort_pivot.iloc[:, 0]
retention_matrix = cohort_pivot.divide(cohort_size, axis=0)

with sns.axes_style("white"):
    fig, ax = plt.subplots(1, 2, figsize=(12, 8), sharey=True, gridspec_kw={
                           'width_ratios': [1, 11]})

    # retention matrix
    sns.heatmap(retention_matrix,
                mask=retention_matrix.isnull(),
                annot=True,
                fmt='.0%',
                cmap='RdYlGn',
                ax=ax[1])
    ax[1].set_title('Monthly Cohorts: User Retention', fontsize=16)
    ax[1].set(xlabel='# of periods',
              ylabel='')

    # cohort size
    cohort_size_df = pd.DataFrame(cohort_size).rename(
        columns={0: 'cohort_size'})
    white_cmap = mcolors.ListedColormap(['white'])
    sns.heatmap(cohort_size_df,
                annot=True,
                cbar=False,
                fmt='g',
                cmap=white_cmap,
                ax=ax[0])

    fig.tight_layout()